### DISCUSSION TITLE
Creating solidjs lib (lerna monorepo, vite)

### DISCUSSION THREAD

**Discussion by aleksanderd** [solidjs/solid#589] [Q&A] [answered]
Hi!

I have a monorepo with several packages: some components lib and web-app(vite) using it.
And, there is different errors(

if compile lib with rollup for both cjs and esm(tsc only for types) the error is:
```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module...
```

if enable `noExternal` at vite.config.js for the lib, the error is:
```
TypeError: __vite_ssr_import_0__.template is not a function
```

Eventually, I researched app-solid-router and try to use same build config: using tsc for generating jsx and rollup for cjs.
Also, added to package.json:
```
  "type": "module",
  "exports": {
    ".": {
      "solid": "./dist/index.jsx",
      "default": "./dist/index.js"
    }
  },
```

Now the error (in browser console):
```
Uncaught TypeError: template2 is undefined
    getNextElement http://localhost:3000/node_modules/.vite/chunk-LFZYSKLS.js?v=8ba36e6d:359
    get children http://localhost:3000/src/root.tsx:21
   ...
```

I pretty confused while trying different configs( Could you help with it?

How can we build the libs and use them in vite config? Must we use 'noExternal' always?

Thank you!

**Comment by ryansolid** [comment, marked as answer]
Sounds like the client build of Solid is getting in somewhere on the server.. there is no `template` on the server. I wonder if it is not externalized in one of the libraries and getting pulled in twice (hence template2). I believe our plugin should be setting up the server build properly so Im a little bit surprised. I haven't had this problem with Solid Start.

**Reply by aleksanderd** [reply]
Yes it is about "pulled in twice" since incorrect nohoist config:
It is necessary to nohoist solid-start since vite not allow to serve upper dirs and config's `fs.allow = ['../..']` not help for me(
So, at some point of experiments, I added nohoist for all solid deps :)

But now, I got the really matter thing is conditional export config entry in package.json, right?

So, now seems to be work ok, thank you very much!
